### ISSUE TITLE
Resource out of sync with cluster state when retrying failed handlers

### ISSUE THREAD

**Issue by anekdoti** [buehler/dotnet-operator-sdk#298]
**Describe the bug**
When a handler (say, create) is retried, I would expect the resource to be in the state that is stored in the cluster.
Instead, the handler is retried with the resource how it was left from the last attempt to handle.

**To Reproduce**
Implement an operator with the following logic in `CreatedAsync(T resource)`:
- if the annotation X does not exist in the resource add the annotation X to resource.
- throw an exception on the first call of CreatedAsync
- update resource in the cluster.

In our case, the exception occured during the update to the cluster since our validation webhook was not yet available.
The consequence is that the annotation X is never added to the resource on the cluster as the handler "thinks" that it already is present - although it only exists in the resource provided by the retry, but not in the cluster.

**Expected behavior**
My expectation would have been that a retry is based on the state of the cluster and not some intermediate left-over state from the last retry.

**Additional context**
This was experienced on KubeOps 5.4.1

**Comment by buehler** [comment]
Hey @anekdoti 
Well, this is awkward :-)

This is related to #264. Good catch for v6.

**Comment by anekdoti** [comment]
Do you think there will be actually a fix for that, or would you expect that the API will get even closer to the one of the Operator SDK in this regard, i.e., that you have to fetch the most recent state of your entity yourself from Kubernetes? 

Maybe in this case one could still offer the cash as a utility class that can be explicitly used to determine the type of change in the event handler?

**Comment by buehler** [comment]
I guess I should fix this issue. It is definitely an issue. However, depending on the mechanism of the "new" cache (if any) this behavior may change.

**Comment by erin-allison** [comment]
@buehler It looks like the code as written _should be_ reloading the resource from the cluster upon processing a delayed event.

https://github.com/buehler/dotnet-operator-sdk/blob/66c86ac1d801b3c6fda91ecbc9225a0a7244ec85/src/KubeOps/Operator/Controller/ManagedResourceController%7BTEntity%7D.cs#L84-L98

I'm baffled that this isn't working as I have definitely encountered the case where the resource in cluster is removed during the requeue interval, and the message `@"Resource ""{kind}/{name}"" for enqueued event was not present anymore.",` has been logged.

**Comment by buehler** [comment]
hmm... maybe there's another issue then :-(

**Comment by anekdoti** [comment]
My apologies for the confusion! It seems that I overlooked the fetching of the resource from Kubernetes in UpdateResourceData :-( 

In my operator, I add an annotation to the managed resource when it is initialized. Reconciliation of dependent resources only happens after initialization is done. The initialization also adds a finalizer to the resource and the update to the resource performed in RegisterFinalizerAsync is used to complete the initialization, i.e., update the changes to the resource performed during the initialization in the cluster. I observed that sometimes, resources were not initialized and thus never entered the reconciliation. And this seemed to coincide with error messages telling that the update in RegisterFinalizerAsync failed because the validation web hook for the resource timed out. 

It also seemed that this error tended to happen soon after restarting the operator (after updating the image in the deployment). 

One strange observation I could not make sense of: I noted that after deleting the deployment of the operator and recreating it, this problem did not happen at all.
